samcli/cli/types.py

```diff
@@ -9,15 +9,47 @@
 import click
 
 
-def _value_regex(delim):
-    return f'(\\"(?:\\\\.|[^\\"\\\\]+)*\\"|(?:\\\\.|[^{delim}\\"\\\\]+)+)'
+def _match_regex(match, delim):
+    return f'(\\"(?:\\\\{match}|[^\\"\\\\]+)*\\"|(?:\\\\{match}|[^{delim}\\"\\\\]+)+)'
 
 
 KEY_REGEX = '([A-Za-z0-9\\"]+)'
+# Tags have additional constraints and they allow "+ - = . _ : / @" apart from alpha-numerics.
+TAG_REGEX = '[A-Za-z0-9\\"_:\\.\\/\\+-\\@=]'
+
 # Use this regex when you have space as delimiter Ex: "KeyName1=string KeyName2=string"
-VALUE_REGEX_SPACE_DELIM = _value_regex(" ")
+VALUE_REGEX_SPACE_DELIM = _match_regex(match=".", delim=" ")
 # Use this regex when you have comma as delimiter Ex: "KeyName1=string,KeyName2=string"
-VALUE_REGEX_COMMA_DELIM = _value_regex(",")
+VALUE_REGEX_COMMA_DELIM = _match_regex(match=".", delim=",")
+
+
+def _unquote(value):
+    r"""
+    Removes wrapping double quotes and any '\ ' characters. They are usually added to preserve spaces when passing
+    value thru shell.
+
+    Examples
+    --------
+    >>> _unquote('val\ ue')
+    value
+
+    >>> _unquote("hel\ lo")
+    hello
+
+    Parameters
+    ----------
+    value : str
+        Input to unquote
+
+    Returns
+    -------
+    Unquoted string
+    """
+    if value and (value[0] == value[-1] == '"'):
+        # Remove quotes only if the string is wrapped in quotes
+        value = value.strip('"')
```

Comment:
The intention here is to only remove the first and last quotes right? Do we need to worry about the case where there are two `"` in the beginning or end of the string input?

Below was run in Python3.8
```
>>> value = '"a\""'
>>> value.strip('"')
'a'
>>> value = '""a""'
>>> value.strip('"')
'a'
```

Doing a slice might be better here:
```
>>> value = '"a\""'
>>> value[1:-1]
'a"'
>>> value = '""a""'
>>> value[1:-1]
'"a"'
```

Reply:
This only cares about wrapping quotes and not the entire nesting, can change the function name to reflect that.